✅ 서브웨이 매장 시뮬레이션

Jinyshin opened the pull request:
### [요구사항 정의서](https://www.notion.so/adapterz/PRD-26b394a48061814c9f6ec1ca90451638) 참고 부탁드립니다.

Jinyshin committed:
✅ 서브웨이 매장 시뮬레이션

kiwinam commented on CLI-Program/src/main/java/org/example/data/MenuData.java:
```diff
@@ -0,0 +1,121 @@
+    
+    // 랜덤 선택을 위한 메서드
+    public static String getRandomSandwich() {
+        return SANDWICH_TYPES.get((int) (Math.random() * SANDWICH_TYPES.size()));
```
랜덤한 메뉴를 가져오는 것을 위해서
Math.random() 과 각 타입의 size 를 곱하고, 인티져로 만들어서 가져오는 코드가 반복됩니다.

이 걸 하나의 메서드로 만들어 전달된 매개변수의 타입으로 랜덤 값을 생성하게 해보는건 어떨까요?
정답이 있는건 아니지만, 코드에서의 중복되는 부분을 의도적으로 하나의 메서드 블럭으로 만들어보는 것도 좋습니다!

kiwinam commented on CLI-Program/src/main/java/org/example/thread/CustomerThread.java:
```diff
@@ -0,0 +1,109 @@
+    while (!Thread.currentThread().isInterrupted()) {
+      List<FoodItem> item = new ArrayList<>();
+      try {
+        if (threadId == 1) {
```
if-else 에서 조건이 3개 이상 반복될 땐, switch 를 활용해보는 것도 좋습니다. 

이유는 조건문 자체가 3개 이상이라는건 조건이 한 눈에 보이지 않기 때문에, 
조건문에서 조건에 따른 프로세스의 흐름을 이해하기 어려울 수도 있습니다.

kiwinam commented on CLI-Program/src/main/java/org/example/thread/CustomerThread.java:
```diff
@@ -0,0 +1,109 @@
+          String meatType = MenuData.getRandomMeat();
+          String breadType = MenuData.getRandomBread();
+          String cheeseType = MenuData.getRandomCheese();
+          String sauceType = MenuData.getRandomSauce();
```
샌드위치에 정보들을 넣는 코드가 반복되고 있는데요, 

샌드위치를 랜덤으로 만들어주는 샌드위치 클래스에 생성자를 하나 만들어서 사용해보면 어떨까요? 사용자 정의 생성자로
그러면 중복 코드를 줄일 수 있고, 재사용하는 것도 편해보입니다!

kiwinam commented on CLI-Program/src/main/java/org/example/domain/Order.java:
```diff
@@ -0,0 +1,30 @@
+    this.id = nextId.getAndIncrement();
+    this.customerName = customerName;
+    this.items = items;
+    this.status = "PENDING";
```
status 에서 사용하는 스트링은 ENUM 으로 처리해보는건 어떨까요?
다른 곳에서 하드 코딩이 있지만, 여기 status 부분은 변경될 가능성이 적어서 이 부분만 ENUM 처리하면 더 좋을 것 같습니다!

kiwinam commented on CLI-Program/src/main/java/org/example/domain/item/FoodItem.java:
```diff
@@ -0,0 +1,20 @@
+
+  public int getPrice() {
+    return price;
+  }
```
코드에서 음식의 정보들은 변경되지 않는다고 했었지만..

이미 캡슐화 되어 있는 클래스는 set 메서드 없는 상태로 두면 기능 확장 시 대응이 어려워서, 클래스 설계할 때 적어주는게 좋습니다.

set 을 아예 설정하지 않으려면 클래스 설계 시에 이 클래스는 절대 변경이 없을 것이다 라는 전제가 필요합니다..!

그래도 가격이랑 이름이 절대 바뀌지 않는다는 가정이라면 괜찮습니다.

kiwinam commented on CLI-Program/src/main/java/org/example/domain/item/MainItem.java:
```diff
@@ -0,0 +1,14 @@
+package org.example.domain.item;
+
+public class MainItem extends FoodItem {
```
Item 이라는 접미어가 붙어 있는데,
리팩토링 적 관점에서는 큰 의미가 없어 보입니다.

클래스명이나 메서드명, 변수명 들은 의미하는 바를 명확하게 표기하는게 좋습니다.

Item 이라는 것이 명확하지 않은 개념이라서 다음에는 더 명확한 네이밍으로 해보시는 것도 좋습니다!

kiwinam commented on CLI-Program/src/main/java/org/example/thread/EmployeeThread.java:
```diff
@@ -0,0 +1,50 @@
+      try {
+        Order order = store.getNextOrder();
+
+        order.setStatus("IN_PROGRESS");
```
이 부분도 status 같이 타입이 정해져 있는 것들은 ENUM 으로 처리해주시면 더 안정적으로 작업 가능합니다!

kiwinam commented on CLI-Program/src/main/java/org/example/thread/LoggerThread.java:
```diff
@@ -0,0 +1,55 @@
+    } catch (InterruptedException e) {
+      // interrupt 상황에서도 로그를 넣기 위해 offer 사용
+      String timestamp = LocalTime.now().format(timeFormatter);
+      String formattedMessage = "[" + timestamp + "] " + message;
```
로깅 포맷 만들어주는 로직도 함수화 시키는 것도 좋아보여요~!

kiwinam reviewed (approved):
전반적으로 코드 잘 작성해주셨습니다.

다만 코드에 주석이 상세하게 적혀 있지 않아서 코드 작성 후 의도에 대해 잘 파악하기 어려울 수도 있어 보입니다.
사람마다 의견이 다르지만, 제 생각은 주석을 적지 않기 위해서는 코드에서 주석이 필요 없을 정도로 메서드와 로직에 대해 가독성이 높게 적어야 하는데, 그렇게 작성하기가 어렵습니다.
그래서 주석으로 어떠한 흐름인지, 어떤 목적이 있는 메서드인지 적어주시면 더 좋을 것 같습니다.

고생하셨습니다!

Jinyshin committed:
refactor: 네이밍 변경 및 팩토리 메서드 도입

- 클래스명 개선
- Sandwich, Salad 클래스에 랜덤 생성 팩토리 메서드 추가
- OrderStatus enum 도입으로 문자열 리터럴 대체 및 타입 안전성 확보
- MenuData에 공통 메서드 getRandomItem 추출로 코드 중복 제거
- CustomerThread를 switch문과 팩토리 메서드 사용으로 리팩토링
- LoggerThread에 formatLogMessage 메서드 추가로 중복 코드 제거
